### PR TITLE
add ability to refresh jobs

### PR DESCRIPTION
**Pull request by DmytroHryshyn**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -49,6 +49,10 @@
 			}
 		],
 		"commands": [
+			{
+				"command": "intuita.refreshStaleJobs",
```
Comment (grzpab) [inline package.json:53]: maybe make it just `refresh jobs`?

File: src/components/messageBus.ts [hunk 2]
```diff
@@ -112,6 +112,13 @@ export type Command =
 			engine: Engine;
 			storageUri: Uri;
 			uri: Uri;
+	  }>
+	| Readonly<{
+			kind: 'executeCodemod';
```
Comment (grzpab) [inline src/components/messageBus.ts:117]: is that supported by the nora node engine?

File: src/components/jobManager.ts [hunk 3]
```diff
@@ -406,4 +409,81 @@ export class JobManager {
...
+					engine: 'node',
+					storageUri,
+					uris: filePaths,
+				},
+				executionId,
+				happenedAt,
+			}),
+		);
+
+		console.log(messages);
```
Comment (grzpab) [inline src/components/jobManager.ts:446]: can you remove it?

File: src/components/jobManager.ts [hunk 4]
```diff
@@ -406,4 +409,81 @@ export class JobManager {
...
+
+		// for (const message of messages) {
+		// 	this.#messageBus.publish(message);
+		// }
+	}
+
+	public async getStaleJobs(): Promise<Job[]> {
```
Comment (grzpab) [inline src/components/jobManager.ts:453]: shouldn't it be private?

File: src/components/engineService.ts [hunk 5]
```diff
@@ -211,6 +214,30 @@ export class EngineService {
 				return args;
 			}
 
+			// execute single codemod
+			if (
+				'kind' in message.command &&
+				message.command.kind === 'executeCodemod'
+			) {
+				// @TODO engine should support such calls
```
Comment (grzpab) [inline src/components/engineService.ts:222]: so this feature is not implemented yet right?

File: src/components/messageBus.ts (hunk 2 above)
Comment (DmytroHryshyn) [reply src/components/messageBus.ts:117]: not yet,  that is what we need to add to the engine

File: src/components/jobManager.ts (hunk 4 above)
Comment (DmytroHryshyn) [reply src/components/jobManager.ts:453]: I think sometimes we may want to get stale jobs to display them somewhere

File: src/components/engineService.ts (hunk 5 above)
Comment (DmytroHryshyn) [reply src/components/engineService.ts:222]: yes, its not implemented, but i still can work on this feature and connect it to the node engine when its done.

File: src/extension.ts [hunk 6]
```diff
@@ -68,6 +68,11 @@ import { randomBytes } from 'crypto';
 const messageBus = new MessageBus();
 
 export async function activate(context: vscode.ExtensionContext) {
+
+	if(!context.storageUri) {
+		throw new Error('Storage Uri not found');
```
Comment (grzpab) [inline src/extension.ts:73]: when can this happen?

File: src/components/jobManager.ts [hunk 7]
```diff
@@ -327,4 +333,45 @@ export class JobManager {
...
+				codemodHash,
+				engine: 'node',
+				storageUri: this.__storageUri,
+				uri,
+			},
+			executionId,
+			happenedAt,
+		}));
+
+		console.log(messages, 'test');
```
Comment (grzpab) [inline src/components/jobManager.ts:371]: can you remove it?

File: src/components/jobManager.ts [hunk 8]
```diff
@@ -327,4 +333,45 @@ export class JobManager {
...
+			if(jobCodemod === null) {
+				return;
+			} 
+			codemodHashUriTuple.push([jobCodemod.hashDigest, job.oldUri]);
+		});
+
+		const executionId = buildExecutionId();
+		const happenedAt = String(Date.now());
+
+		const messages: Message[] = codemodHashUriTuple.map(([codemodHash, uri]) =>({
```
Comment (grzpab) [inline src/components/jobManager.ts:358]: how does that work with the principle that only one execution can be managed?

Comment (DmytroHryshyn) [reply src/components/jobManager.ts:358]: Currently we can have only 1 execution at a time, and we can only pass 1 codemod to the engine, right? 
I will need to add ability to `await` for execution to complete and then chain executions. 
Something like 
```
for(const codemod of codemods) {
await this.__engineService.executeCodemod(codemod, uris);
}
```

File: src/extension.ts (hunk 6 above)
Comment (DmytroHryshyn) [reply src/extension.ts:73]: storageUri has type `uri | undefined`, but im not sure in what cases it can be undefined.

**Pull request closed without merging**